Add optional publisher for events

jcredding opened the pull request:
This adds an optional publisher that can either be configured
globally or specified when publishing an event. This can be used
for tracking purposes and identifying where an event was published
from for debugging. There is now a `publish_as` method which can
be used to specify a publisher when publishing an event. This
method works the same as the `publish` method only it sets the
publisher.

@kellyredding - Ready for review.

kellyredding commented:
@jcredding this sparked a number of discussions yesterday evening.  There are now some other stuff that needs to happen before we progress with this (payload {de}serialize{_hash}, events/jobs subclassing a common message object, etc).

Specifically to this effort, we discussed the message objs taking a single options hash with `:params` being one of the options.  Top `Qs` enqueue ans publish methods don't take options hash.  Add `publish_as` method that takes the publisher as the first arg then the publish args following it.

We'll relook at this PR once all that has been addressed.

jcredding commented:
@kellyredding - I've rebased and updated this. There is now a `publish_as` method that takes a custom publisher and the `publish` and `enqueue` methods no longer take options.

kellyredding commented on test/unit/client_tests.rb:
```diff
@@ -105,6 +105,24 @@ class MixinTests < UnitTests
+      })
+      assert_equal dispatch_job.name,   call.job.name
+      assert_equal dispatch_job.params, call.job.params
+    end
```
@jcredding I'd add tests for returning the event and also enqueueing on the dispatch queue like you do in the publish method tests above.

kellyredding commented:
@jcredding cool man.  One minor comment about the publish as method test.  Otherwise :-1: 

I'm excited to see the single opts arg PR coming next!  I think it will clean things up nicely.
